File: Grind75/32-CloneGraph/cpp/CloneGraph.cpp
#include "CloneGraph.h"
#include <iostream>
#include <queue>
#include <unordered_map>
#include <utility>

using namespace std;
typedef GraphNode Node;

Node* helper(Node *node, unordered_map<Node*, Node*>& cloned) {
  if (cloned.find(node) != cloned.end())
    return cloned[node];
  Node * clone = new Node(node->val);
  cloned[node] = clone;
  for (auto neighbor : node->neighbors)
    clone->neighbors.push_back(helper(neighbor, cloned));
  return cloned[node];
}
Node* CloneGraph::cloneGraph_recursive(Node* node) {
  if (!node) return nullptr;
  unordered_map<Node*, Node*> clones;
  return helper(node, clones);
}

Node* CloneGraph::cloneGraph(Node* node) {
  if (!node) return nullptr;

  Node *start = new Node(node->val);
  unordered_map<Node*, Node*> cloned;
  cloned[node] = start;

  queue<Node*> q;
  q.push(node);

  while (!q.empty()) {
    Node *node = q.front();
    q.pop();
    for (auto neighbor : node->neighbors) {
      if (cloned.find(neighbor) == cloned.end()) {
        cloned[neighbor] = new Node(neighbor->val);
        q.push(neighbor);
      }
      cloned[node]->neighbors.push_back(cloned[neighbor]);
    }
  }
  return start;
}

File: Grind75/32-CloneGraph/python/CloneGraph.py
#!/bin/python3

from __future__ import annotations
import sys
sys.path.append('../../Utility/python')

from typing import Optional, Dict, Deque
from collections import deque

class Node:
  def __init__(self, val = 0, neighbors = None):
    self.val = val
    self.neighbors = neighbors if neighbors is not None else []

class Solution(object):
  def cloneGraph(self, node: Optional[Node]) -> Optional[Node]:
    if node is None:
      return None
    q: Deque[Node] = deque()
    q.append(node)
    first: Node = Node(node.val)
    cloned: Dict[Node, Node] = dict()
    cloned[node] = first

    while len(q) > 0:
      node = q.pop()
      for neighbor in node.neighbors:
        if neighbor not in cloned:
          cloned[neighbor] = Node(neighbor.val)
          q.appendleft(neighbor)
        cloned[node].neighbors.append(cloned[neighbor])
    return first

if __name__ == '__main__':
  soln = Solution()
  node1 = Node(1)
  node2 = Node(2)
  node3 = Node(3)
  node4 = Node(4)
  node1.neighbors = [node2, node4]
  node2.neighbors = [node1, node3]
  node3.neighbors = [node2, node4]
  node4.neighbors = [node1, node3]

  newGraph = soln.cloneGraph(node1)

